Listener: extend connection balancer (#20268)
envoyproxy/envoy: docs/root/configuration/other_features/dlb.rst

.. _config_connection_balance_dlb:

> Dlb Connection Balancer

Reviewer: There should be a release note for this.

Author: Added.

.. _config_connection_balance_dlb:

> DLB Connection Balancer

Reviewer 2: ```suggestion
DLB connection balancer
```

@daixiang0 here you use `DLB` whereas elsewhere you use `Dlb` - i think we want to make this consistent

Author: Let me change it to `Dlb`.

=======================

* :ref:`v3 API reference <envoy_v3_api_msg_extensions.dlb.v3alpha.Dlb>`

This connection balancer extension provides Envoy with low latency networking by integrating with `Intel DLB <https://networkbuilders.intel.com/solutionslibrary/queue-management-and-load-balancing-on-intel-architecture>`_ through the libdlb library.

The DLB connection balancer is only included in :ref:`contrib images <install_contrib>`

Example configuration
---------------------

> .. code-block:: yaml

Reviewer 2: please use a `literalinclude` and separate yaml file for this 

(you may need to exclude archs/platforms i guess)

Author: Is there an example for excluding archs/platforms?

  static_resources:
    listeners:
      connection_balance_config:
        extend_balance:
          name: envoy.network.connection_balance.dlb
          typed_config:
            "@type": type.googleapis.com/envoy.extensions.dlb.v3alpha.Dlb

How it works
------------

> If enabled, the DLB connection balancer will

Reviewer 2: ```suggestion
If enabled, the DLB connection balancer will:
```

Author: Done.

> - attache DLB hardware

Reviewer 2: this list is collapsing - i think it needs a space above

Reviewer 2: ```suggestion
- attach DLB hardware
```

Author: Sure.

- create a queue for balancing
- create one port to send and one port to receive for each worker thread
- create one eventfd for each worker thread and attach each eventfd to corresponding customer
- register each eventfd to corresponding customer and DLB hardware

> When new connections come, one worker thread will accept it and send it to DLB hardware. DLB hardware
> does balancing then trigger one worker thread to receive via libevent.

Reviewer 2: ```suggestion
When a new connection is established, a single worker thread accepts it notifies the DLB hardware. The DLB hardware performs the load balancing and in turn triggers a worker thread via libevent.
```

not sure if my suggestion is accurate - i didnt fully understand the original sentence

Author: I mean:
> When a new connection is established, a single worker thread accepts it and send its' pointer to the DLB hardware. The DLB hardware performs the load balancing and in turn triggers a worker thread via libevent.

Is it more clear?

Installing and using DLB
------------------------

> For information on how to build/install and use libdlb see the getting started guide `here <https://downloadmirror.intel.com/727424/DLB_Driver_User_Guide.pdf>`_.

Reviewer 2: ```suggestion
For information on how to build/install and use libdlb see the `getting started guide <https://downloadmirror.intel.com/727424/DLB_Driver_User_Guide.pdf>`_.
```

Author: Done.
